homeassistant/components/colorthief/__init__.py

```diff
@@ -0,0 +1,142 @@
+"""Module for ColorThief (RGB extraction from images) component."""
+import asyncio
+import io
+import logging
+
+from PIL import UnidentifiedImageError
+import aiohttp
+import async_timeout
+from colorthief import ColorThief
+
+from homeassistant.components.colorthief.const import (
+    ATTR_FILE_PATH,
+    ATTR_LIGHT_ENTITY_ID,
+    ATTR_URL,
+    DOMAIN,
+    SERVICE_PREDOMINANT_COLOR_FILE,
+    SERVICE_PREDOMINANT_COLOR_URL,
+)
+from homeassistant.components.light import (
+    ATTR_BRIGHTNESS_PCT,
+    ATTR_RGB_COLOR,
+    ATTR_TRANSITION,
+    DOMAIN as LIGHT_DOMAIN,
+    SERVICE_TURN_ON,
+)
+from homeassistant.const import ATTR_ENTITY_ID
+from homeassistant.helpers import aiohttp_client
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def _get_file(file_path):
+    return file_path
+
+
+async def async_setup(hass, hass_config):
+    """Set up services for ColorThief integration."""
+
+    _LOGGER.debug("Setting up ColorThief component")
+
+    async def _async_get_color(file_handler) -> tuple:
+        """Given an image file, extract the predominant color from it."""
+        try:
+            color_thief = await hass.async_add_executor_job(ColorThief, file_handler)
+        except UnidentifiedImageError as ex:
+            _LOGGER.error("Bad image file provided, are you sure it's an image? %s", ex)
+            return
+
+        # get_color returns a SINGLE RGB value for the given image
+        color = color_thief.get_color(quality=1)
+
+        _LOGGER.debug("Extracted RGB color %s from image", color)
+
+        return color
+
+    async def _async_set_light(
+        light_entity_id, color, brightness_pct=None, transition=None
+    ):
+        """Set the given light to our extracted RGB value."""
+        service_data = {
+            ATTR_ENTITY_ID: light_entity_id,
+            ATTR_RGB_COLOR: color,
+        }
+
+        if brightness_pct:
+            service_data[ATTR_BRIGHTNESS_PCT] = brightness_pct
+
+        if transition:
+            service_data[ATTR_TRANSITION] = transition
+
+        _LOGGER.debug("Setting RGB %s on light %s", color, light_entity_id)
+
+        await hass.services.async_call(
+            LIGHT_DOMAIN, SERVICE_TURN_ON, service_data, blocking=True
+        )
+
+    async def async_predominant_color_url_service(service_call):
+        """Handle call for URL based image."""
+        service_data = service_call.data
+
+        url = service_data.get(ATTR_URL)
```

Comment:
We should check if the url is allowed `hass.config.is_allowed_external_url(url)`

Reply:
Is there a nicer way to test the homeassistant config for allowlist_* settings? Currently mocking the method to return True (copied from another test) but feel it would be nicer not to mock that and just set it somewhere during the test.

Reply:
`hass.config.allowlist_external_dirs` is a set. You can call `.add("path")` on it.

Reply:
Very nice, I've added both checks correctly now